framework/bundles/org.jboss.tools.rsp.secure/src/org/jboss/tools/rsp/secure/model/RSPSecureStorage.java

```diff
@@ -0,0 +1,169 @@
+	
+	
+	private File backingFile;
+	private byte[] key;
+	
+	private Properties secretData;
+	private CryptoUtils util;
+	
+	public RSPSecureStorage(File file, byte[] key) {
+		this.backingFile = file;
+		this.util = new CryptoUtils();
+		this.key = util.keyTo16(key);
+	}
+	
+	public void setKey(byte[] key) {
+		this.key = key;
+		secretData = null;
+	}
+	
+	public void load() throws CryptoException {
+		Properties tmp = new Properties();
+		if( backingFile != null && backingFile.exists()) {
+			try {
+				byte[] encrypted = util.getBytesFromFile(backingFile);
+				byte[] decrypted = util.decrypt(key, encrypted);
+				
+				byte[] magicBytes = HASH_COMMENT.getBytes(); 
+				if( !startsWith(decrypted, magicBytes)) {
+					throw new CryptoException("Invalid key", null);
+				}
+				tmp.load(new ByteArrayInputStream(decrypted));
+				Arrays.fill(decrypted, (byte)0);
+				secretData = tmp;
+			} catch(IOException | CryptoException e) {
+				throw new CryptoException("Unable to decrypt secure storage", e);
+			}
+		} else if( !backingFile.exists()) {
+			// new file, it's ok to create
+			secretData = tmp;
+		}
+	}
+	
+	private boolean startsWith(byte[] all, byte[] magic) {
+		for( int i = 0; i < magic.length; i++ ) {
+			if( i >= all.length) {
+				return false;
+			}
+			if( all[i] != magic[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+
+	private boolean isInitialized() {
+		return secretData != null;
+	}
+	
+	public void save() throws CryptoException {
+		try {
+			if( backingFile != null && isInitialized()) {
+				backingFile.getParentFile().mkdirs();
+				StringWriter sw = new StringWriter();
+				secretData.store(sw, COMMENT);
+				byte[] raw = sw.toString().getBytes();
+				byte[] encrypted = util.encrypt(key, raw);
+				util.writeBytesToFile(backingFile, encrypted);
+			} else if( !isInitialized()) {
+				throw new NotInitializedCryptoException("Secure storage not initialized", null);
+			}
+		} catch(IOException ioe) {
+			throw new CryptoException("Secure storage not initialized", ioe);
+		}
+	}
+
+	@Override
+	public boolean propertyExists(String nodePath, String propertyName) throws NotInitializedCryptoException {
+		if( !isInitialized()) {
+			throw new NotInitializedCryptoException("Secure storage not initialized", null);
+		}
+		return ((SecureNode)getNode(nodePath)).propertyExists(propertyName);
+	}
+
+	public ISecureNode getNode(String nodePath) throws NotInitializedCryptoException {
+		if( !isInitialized()) {
+			throw new NotInitializedCryptoException("Secure storage not initialized", null);
+		}
+		return new SecureNode(nodePath);
+	}
+	
+	public class SecureNode implements ISecureNode {
+		private String path;
+		public SecureNode(String path) {
+			this.path = path;
+		}
+		
+		public SecureNode getChildNode(String segment) {
+			return new SecureNode(append(segment));
+		}
+		
+		private String append(String segment) {
+			String toCheck = path;
+			if( !path.endsWith("/"))
+				toCheck = toCheck + "/";
```

Comment:
i think `File.pathSeparator` is better here to avoid problems with Windows

Reply:
This is a string inside a file. Windows will not have a problem with a text file that has a slash in it.

Reply:
@bdshadow I misunderstood it too. The separator is within the keys, it's about being able to have some sort of tree-like structure (like the one in Eclipse secure storage). It's not a path separator.